PSReadLine/Render.Helper.cs

```diff
@@ -0,0 +1,161 @@
+--********************************************************************/
+
+using System;
+
+namespace Microsoft.PowerShell
+{
+    public partial class PSConsoleReadLine
+    {
+        private void WriteBlankLines(int count)
+        {
+            _console.BlankRestOfLine();
+            for (int i = 1; i < count; i++)
+            {
+                _console.Write("\n");
+                _console.BlankRestOfLine();
+            }
+        }
+
+        private void WriteBlankLines(int top, int count)
+        {
+            var savedCursorLeft = _console.CursorLeft;
+            var savedCursorTop = _console.CursorTop;
+
+            _console.SetCursorPosition(left: 0, top);
+            WriteBlankLines(count);
+            _console.SetCursorPosition(savedCursorLeft, savedCursorTop);
+        }
+
+        private void WriteBlankRestOfLine(int left, int top)
+        {
+            var savedCursorLeft = _console.CursorLeft;
+            var savedCursorTop = _console.CursorTop;
+
+            _console.SetCursorPosition(left, top);
+            _console.BlankRestOfLine();
+            _console.SetCursorPosition(savedCursorLeft, savedCursorTop);
+        }
+
+        private static string Spaces(int cnt)
+        {
+            return cnt < _spaces.Length
+                ? (_spaces[cnt] ?? (_spaces[cnt] = new string(' ', cnt)))
+                : new string(' ', cnt);
+        }
+
+        private static int LengthInBufferCells(string str)
+        {
+            return LengthInBufferCells(str, 0, str.Length);
+        }
+
+        private static int LengthInBufferCells(string str, int start, int end)
+        {
+            var sum = 0;
+            for (var i = start; i < end; i++)
+            {
+                var c = str[i];
+                if (c == 0x1b && (i+1) < end && str[i+1] == '[')
+                {
+                    // Simple escape sequence skipping
+                    i += 2;
+                    while (i < end && str[i] != 'm')
+                        i++;
+
+                    continue;
+                }
+                sum += LengthInBufferCells(c);
+            }
+            return sum;
+        }
+
+        private static int LengthInBufferCells(char c)
+        {
+            if (c < 256)
+            {
+                // We render ^C for Ctrl+C, so return 2 for control characters
+                return Char.IsControl(c) ? 2 : 1;
+            }
+
+            // The following is based on http://www.cl.cam.ac.uk/~mgk25/c/wcwidth.c
+            // which is derived from http://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt
+
+            bool isWide = c >= 0x1100 &&
+                (c <= 0x115f || /* Hangul Jamo init. consonants */
+                    c == 0x2329 || c == 0x232a ||
+                    (c >= 0x2e80 && c <= 0xa4cf &&
+                    c != 0x303f) || /* CJK ... Yi */
+                    (c >= 0xac00 && c <= 0xd7a3) || /* Hangul Syllables */
```

Comment:
Is performance important here?
```suggestion
                    (((uint)c - (uint)0xac00) <= ((uint)0xd7a3) - (uint)0xac00) || /* Hangul Syllables */
```
We could use helper method https://source.dot.net/#System.Private.CoreLib/UnicodeUtility.cs,141

Reply:
This is existing code and I'd like to not change it.
How much is the `(value - lower) <= (upper - lower)` faster than `value >= lower && value <= upper`?

Reply:
This patterns comes from .Net repo. They uses it broadly in all char operations. It use 2 operation instead of 3 ones in first pattern and more fast commands. It is huge optimization relatively (up to 30%) and some ns in absolute values.

Another way they uses is 2- or 3-level translation tables.

Reply:
Interesting, let me open a new issue tracking the investigation to see if we want to replace them with the .NET pattern.
Issue opened: #1915